### PR TITLE
[extension/encoding/jsonlogencoding] add new extension for JSON codec

### PR DESCRIPTION
**Pull request by VihasMakwana**
**Description:** Create a new extension for JSON. This will be used in pulsarreceiver/kafkareceiver to populate the log record's map from the raw body.

**Link to tracking Issue:** [6272](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/28686)

**Testing:** Relevant cases added

### CODE REVIEW

**Comment by VihasMakwana** [comment]
@MovieStoreGuy @atoulme @dmitryax I would appreciate your review as well.

**Comment by VihasMakwana** [comment]
@MovieStoreGuy for some reasons I cant update CODEOWNERS (maybe because I'm not member yet?).
I'm adding you as a codeowner as well, is that fine for you?

**Comment by dmitryax** [comment]
> maybe because I'm not member yet?

That's true. I suggest you to apply for membership give that you contributed enough already. Find two sponsors (from Splunk and outside) and submit a PR similar to https://github.com/open-telemetry/community/issues/1714

**Comment by dmitryax** [comment]
@atoulme you might be also interested to be a code owner?

File: extension/encoding/jsonencodingextension/metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+type: jsonencoding
```
Comment (dmitryax) [inline extension/encoding/jsonencodingextension/metadata.yaml:1]:
```suggestion
type: json_log_encoding
```

Also let's call the go module jsonlogencodingextension

File: extension/encoding/jsonencodingextension/extension.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+var _ plog.Unmarshaler = &jsonExtension{}
+var _ plog.Marshaler = &jsonExtension{}
+
+type jsonExtension struct {
+}
+
+func (e *jsonExtension) MarshalLogs(ld plog.Logs) ([]byte, error) {
+	marshaler := &plog.JSONMarshaler{}
+	return marshaler.MarshalLogs(ld)
```
Comment (dmitryax) [inline extension/encoding/jsonencodingextension/extension.go:24]: The marshaling doesn't match unmarshaling. Please fix it and add a test covering marshal->unmarshal sequence and validating that the result matches the original logs

File: extension/encoding/jsonencodingextension/extension.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+)
+
+var _ plog.Unmarshaler = &jsonExtension{}
+var _ plog.Marshaler = &jsonExtension{}
```
Comment (dmitryax) [inline extension/encoding/jsonencodingextension/extension.go:17]: These should not be used in this module. They are part of the OTLP encoder

**Comment by atoulme** [comment]
I can be codeowner.

File: extension/encoding/jsonencodingextension/metadata.yaml (hunk 1 above)
Comment (astencel-sumo) [reply extension/encoding/jsonencodingextension/metadata.yaml]: Why are we limiting this encoding extension to logs only? Wouldn't it make sense to have a single extension that (in future) supports encoding of all signals - logs, metrics, traces? Similar to how we don't have separate `otlplogs` receiver, but a single `otlp` receiver for logs, metrics and traces.

Comment (astencel-sumo) [reply extension/encoding/jsonencodingextension/metadata.yaml]: Oh I think I misunderstood the purpose of this encoding extension. I assumed it is an "OTLP JSON" encoding extension, whereas it is an extension to encode/decode JSON log body, right?

**Comment by VihasMakwana** [comment]
@dmitryax @astencel-sumo can you review this one?

**Comment by atoulme** [comment]
PTAL at the failing tests?

File: extension/encoding/jsonlogencodingextension/json_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	assert.NoError(t, err)
+	assert.Equal(t, 1, ld.LogRecordCount())
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		for j := 0; j < ld.ResourceLogs().At(i).ScopeLogs().Len(); j++ {
+			for k := 0; k < ld.ResourceLogs().At(i).ScopeLogs().At(j).LogRecords().Len(); k++ {
+				lr := ld.ResourceLogs().At(i).ScopeLogs().At(j).LogRecords().At(k)
+				lr.SetObservedTimestamp(pcommon.NewTimestampFromTime(time.Unix(0, 0)))
+			}
+		}
+	}
```
Comment (dmitryax) [inline extension/encoding/jsonlogencodingextension/json_test.go:29]: I don't think this is needed. We don't use the observed timestamp in marshalling

File: extension/encoding/jsonlogencodingextension/README.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+| Issues        | [![Open issues](https://img.shields.io/github/issues-search/open-telemetry/opentelemetry-collector-contrib?query=is%3Aissue%20is%3Aopen%20label%3Aextension%2Fjsonlogencoding%20&label=open&color=orange&logo=opentelemetry)](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues?q=is%3Aopen+is%3Aissue+label%3Aextension%2Fjsonlogencoding) [![Closed issues](https://img.shields.io/github/issues-search/open-telemetry/opentelemetry-collector-contrib?query=is%3Aissue%20is%3Aclosed%20label%3Aextension%2Fjsonlogencoding%20&label=closed&color=blue&logo=opentelemetry)](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues?q=is%3Aclosed+is%3Aissue+label%3Aextension%2Fjsonlogencoding) |
+| [Code Owners](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#becoming-a-code-owner)    | [@VihasMakwana](https://www.github.com/VihasMakwana), [@atoulme](https://www.github.com/atoulme) |
+
+[development]: https://github.com/open-telemetry/opentelemetry-collector#development
+<!-- end autogenerated section -->
+
+The `jsonlog` encoding extention is used to marshal/unmarshal JSON log body.
```
Comment (dmitryax) [inline extension/encoding/jsonlogencodingextension/README.md:14]:
```suggestion
The `jsonlog` encoding extension is used to marshal/unmarshal JSON log body, ignoring other log fields.
```

File: extension/encoding/jsonlogencodingextension/extension.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	"fmt"
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+)
+
+type jsonExtension struct {
```
Comment (dmitryax) [inline extension/encoding/jsonlogencodingextension/extension.go:17]:
```suggestion
type jsonLogEncoding struct {
```

**Comment by VihasMakwana** [comment]
> PTAL at the failing tests?

yeah, running `make update-otel` resolved it and CI's green

**Review by dmitryax — APPROVED** [review]

**Pull request merged**